lib/websocket/connection.js

```diff
@@ -0,0 +1,314 @@
+      //    and this header field indicates the use of a subprotocol that was
+      //    not present in the client's handshake (the server has indicated a
+      //    subprotocol not requested by the client), the client MUST _Fail
+      //    the WebSocket Connection_.
+      const secProtocol = response.headersList.get('Sec-WebSocket-Protocol')
+
+      if (secProtocol !== null && secProtocol !== request.headersList.get('Sec-WebSocket-Protocol')) {
+        failWebsocketConnection(ws)
+        return
+      }
+
+      // processResponse is called when the "response’s header list has been received and initialized."
+      // once this happens, the connection is open
+      ws[kResponse] = response
+
+      const parser = new ByteParser(ws)
+      response.socket.ws = ws // TODO: use symbol
+      ws[kByteParser] = parser
+
+      whenConnectionEstablished(ws)
+
+      response.socket.on('data', onSocketData)
+      parser.on('drain', onParserDrain)
+
+      socketClosed(ws)
+    }
+  })
+
+  return controller
+}
+
+/**
+ * @see https://websockets.spec.whatwg.org/#feedback-from-the-protocol
+ * @param {import('./websocket').WebSocket} ws
+ */
+function whenConnectionEstablished (ws) {
+  const { [kResponse]: response } = ws
+
+  // 1. Change the ready state to OPEN (1).
+  ws[kReadyState] = states.OPEN
+
+  // 2. Change the extensions attribute’s value to the extensions in use, if
+  //    it is not the null value.
+  // https://datatracker.ietf.org/doc/html/rfc6455#section-9.1
+  const extensions = response.headersList.get('sec-websocket-extensions')
+
+  if (extensions !== null) {
+    ws[kExtensions] = extensions
+  }
+
+  // 3. Change the protocol attribute’s value to the subprotocol in use, if
+  //    it is not the null value.
+  // https://datatracker.ietf.org/doc/html/rfc6455#section-1.9
+  const protocol = response.headersList.get('sec-websocket-protocol')
+
+  if (protocol !== null) {
+    ws[kProtocol] = protocol
+  }
+
+  // 4. Fire an event named open at the WebSocket object.
+  fireEvent('open', ws)
+
+  if (channels.open.hasSubscribers) {
+    channels.open.publish({
+      address: response.socket.address(),
+      protocol,
+      extensions
+    })
+  }
+}
+
+/**
+ * @param {Buffer} chunk
+ */
+function onSocketData (chunk) {
+  if (!this.ws[kByteParser].write(chunk)) {
+    this.pause()
+  }
+}
+
+function onParserDrain () {
+  this.ws[kResponse].socket.resume()
+}
+
+/**
+ * @see https://websockets.spec.whatwg.org/#feedback-from-the-protocol
+ * @see https://datatracker.ietf.org/doc/html/rfc6455#section-7.1.4
+ * @param {import('./websocket').WebSocket} ws
+ */
+function socketClosed (ws) {
+  const { [kResponse]: response } = ws
+
+  response.socket.on('close', () => {
+    // If the TCP connection was closed after the
+    // WebSocket closing handshake was completed, the WebSocket connection
+    // is said to have been closed _cleanly_.
+    const wasClean = ws[kSentClose]
```

Comment:
A sent close frame is not sufficient. A close frame should also be received. It should be `ws[kSentClose] && ws[kReceivedClose]` assuming that `ws[kReceivedClose]` is set when a close frame is received.